loopdb/store.go

```diff
@@ -318,44 +330,56 @@ func (s *boltSwapStore) createLoop(bucketKey []byte, hash lntypes.Hash,
 	})
 }
 
-// CreateLoopOut adds an initiated swap to the store.
+// CreateLoopIn adds an initiated swap to the store.
 //
 // NOTE: Part of the loopdb.SwapStore interface.
-func (s *boltSwapStore) CreateLoopOut(hash lntypes.Hash,
-	swap *LoopOutContract) error {
+func (s *boltSwapStore) CreateLoopIn(hash lntypes.Hash,
+	swap *LoopInContract) error {
 
 	// If the hash doesn't match the pre-image, then this is an invalid
 	// swap so we'll bail out early.
 	if hash != swap.Preimage.Hash() {
 		return errors.New("hash and preimage do not match")
 	}
 
-	contractBytes, err := serializeLoopOutContract(swap)
-	if err != nil {
-		return err
-	}
+	// Otherwise, we'll create a new swap within the database.
+	return s.db.Update(func(tx *bbolt.Tx) error {
+		// First, we'll grab the root bucket that houses all of our
+		// main swaps.
+		rootBucket, err := tx.CreateBucketIfNotExists(loopInBucketKey)
+		if err != nil {
+			return err
+		}
 
-	return s.createLoop(loopOutBucketKey, hash, contractBytes)
-}
+		// If the swap already exists, then we'll exit as we don't want
+		// to override a swap.
+		if rootBucket.Get(hash[:]) != nil {
+			return fmt.Errorf("swap %v already exists", hash)
+		}
 
-// CreateLoopIn adds an initiated swap to the store.
-//
-// NOTE: Part of the loopdb.SwapStore interface.
-func (s *boltSwapStore) CreateLoopIn(hash lntypes.Hash,
-	swap *LoopInContract) error {
+		// From the root bucket, we'll make a new sub swap bucket using
+		// the swap hash.
+		swapBucket, err := rootBucket.CreateBucket(hash[:])
+		if err != nil {
+			return err
+		}
```

Comment:
Could this all be pulled out into a `createLoopBucket(rootKey)` call to reduce duplication? Or is that the opposite of what we're going for splitting these out.

Reply:
Done. Didn't think about this variant.